pkg/controllers/networkpolicy/tracker.go

```diff
@@ -0,0 +1,283 @@
+	NpStatus map[string]*runtimev1alpha1.NetworkPolicyStatus
+}
+
+// newCloudResourceNpTracker create a tracker object using cloud resource object.
+// Currently, VM is the only cloud resource which has a tracker.
+func (r *NetworkPolicyReconciler) newCloudResourceNpTracker(rsc *cloudresource.CloudResource) *CloudResourceNpTracker {
+	log := r.Log.WithName("NPTracker")
+
+	vmItems, err := r.Inventory.GetVmFromIndexer(indexer.VirtualMachineByCloudResourceID, rsc.CloudResourceID.String())
+	if err != nil || len(vmItems) <= 0 {
+		log.Error(err, "failed to create np tracker for cloud resource id", "id", rsc.String())
+		return nil
+	}
+
+	// A vm is uniquely identified by its cloud assigned id and vpc id.
+	vm := vmItems[0].(*runtimev1alpha1.VirtualMachine)
+	tracker := &CloudResourceNpTracker{
+		appliedToSGs:     make(map[string]*appliedToSecurityGroup),
+		prevAppliedToSGs: make(map[string]*appliedToSecurityGroup),
+		appliedToToNpMap: make(map[string][]types.NamespacedName),
+		CloudResource:    *rsc,
+		// Use vm namespaced name for the tracker.
+		NamespacedName: types.NamespacedName{Name: vm.Name, Namespace: vm.Namespace},
+	}
+
+	log.Info("Create np tracker", "namespacedName", tracker.NamespacedName)
+	if err := r.cloudResourceNpTrackerIndexer.Add(tracker); err != nil {
+		log.Error(err, "add to CloudResourceNpTracker indexer")
+		return nil
+	}
+	return tracker
+}
+
+func (r *NetworkPolicyReconciler) getCloudResourceNpTracker(rsc *cloudresource.CloudResource, create bool) *CloudResourceNpTracker {
+	if obj, found, _ := r.cloudResourceNpTrackerIndexer.GetByKey(rsc.String()); found {
+		return obj.(*CloudResourceNpTracker)
+	} else if create {
+		return r.newCloudResourceNpTracker(rsc)
+	}
+	return nil
+}
+
+// processCloudResourceNpTrackers computes the np status of trackers marked dirty.
+func (r *NetworkPolicyReconciler) processCloudResourceNpTrackers() {
+	log := r.Log.WithName("NPTracker")
+	for _, obj := range r.cloudResourceNpTrackerIndexer.List() {
+		tracker := obj.(*CloudResourceNpTracker)
+		if tracker == nil || !tracker.isDirty() {
+			continue
+		}
+		status := tracker.computeNpStatus(r)
+		if len(tracker.appliedToSGs) == 0 && len(tracker.prevAppliedToSGs) == 0 {
+			log.Info("Delete np tracker", "namespacedName", tracker.NamespacedName)
+			_ = r.cloudResourceNpTrackerIndexer.Delete(tracker)
+			continue
+		}
+
+		npStatus, ok := status[tracker.NamespacedName.Namespace]
+		if !ok {
+			// Should not occur.
+			log.Error(fmt.Errorf("failed to get NetworkPolicy status for tracker"), "", "namespacedName", tracker.NamespacedName)
+			continue
+		}
+		tracker.NpStatus = npStatus
+
+		// After computing the status always update the indexer.
+		if ok {
+			// Remove the log message after debugging.
+			r.Log.V(1).Info("Updating indexer", "namespacedName", tracker.NamespacedName, "NetworkPolicyStatus", tracker.NpStatus)
+			_ = r.cloudResourceNpTrackerIndexer.Delete(tracker)
```

Comment:
Store update as pointer and avoid delete and add. It should work.. You dont unnecessarily need to do this.

Reply:
It is a map.  I tried it and updating the indexer is not required.